docker/DOCKER.md

```diff
@@ -0,0 +1,173 @@
+# Docker Images for balena CLI
+
+Docker images with balena CLI and docker-in-docker.
+
+## Building (required)
+
+Follow these steps to build your own Docker image with
+the CLI and dependencies pre-installed. These steps have been tested
+on x86_64 workstations, building images for other platforms like ARM.
+
+```bash
+# provide the architecture where you will be running the image
+# (see Available Architectures below)
+export BALENA_ARCH="amd64"
+
+# optionally enable QEMU binfmt if building for other architectures (eg. armv7hf)
+docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
+
+# the recommended distro is debian, but alpine Dockerfiles are also available
+export BALENA_DISTRO="debian"
+
+# provde a CLI version from https://github.com/balena-io/balena-cli/releases
+export BALENA_CLI_VERSION="12.40.0"
+
+# build and tag an image with docker
+docker build -f docker/${BALENA_DISTRO}/Dockerfile --build-arg BALENA_ARCH \
+    --tag "balenacli-${BALENA_DISTRO}-${BALENA_ARCH}:${BALENA_CLI_VERSION}" \
+    --tag "balenacli-${BALENA_DISTRO}-${BALENA_ARCH}:latest" \
+    --pull
+```
+
+## Available Architectures
+
+- `rpi`
+- `armv7hf`
+- `aarch64` (debian only)
+- `amd64`
+- `i386`
+
+## Basic Usage
+
+Here's a small example of running a single, detached container
+in the background and using `docker exec` to run balena CLI commands.
+
+```
+$ docker run --detach --privileged --network host --name cli --rm -it balenacli-debian-amd64 /bin/bash
+
+$ docker exec -it cli balena version -a
+balena-cli version "12.38.1"
+Node.js version "12.19.1"
+
+$ docker exec -it cli balena login --token abc...
+
+$ docker exec -it cli balena whoami
+== ACCOUNT INFORMATION
+USERNAME: ...
+EMAIL:    ...
+URL:      balena-cloud.com
+
+$ docker exec -it cli balena apps
+ID      APP NAME         SLUG                            DEVICE TYPE     ONLINE DEVICES DEVICE COUNT
+1491721 test-nuc         gh_paulo_castro/test-nuc        intel-nuc       0              1
+...
+
+$ docker exec -it cli balena app test-nuc
+== test-nuc
+ID:          149...
+DEVICE TYPE: intel-nuc
+SLUG:        gh_.../test-nuc
+COMMIT:      ce9...
+```
+
+## Advanced Usage
+
+The following are examples of running the docker image in various
+modes in order to allow only the required functionality, and not
+elevate permissions unless required.
+
+### scan
+
+- <https://www.balena.io/docs/reference/balena-cli/#scan>
+
+```bash
+# balena scan requires the host network and NET_ADMIN
+docker run --rm -it --cap-add NET_ADMIN --network host \
```

Comment:
In my testing, this was successful on Linux (Debian v10) with "regular Docker" but not on macOS with Docker Desktop. I understand that Docker Desktop uses a virtual machine behind the scenes, and perhaps this is the reason why. Did you get to test with Docker Desktop on macOS or Windows? We may want to document that this is not expected to work with Docker Desktop on macOS or Windows, if you think this sounds right.

Also, on my testing on Linux (Debian v10 and balenaOS), I didn't need `--cap-add NET_ADMIN` for `scan` to work.

Reply:
`NET_ADMIN` isn't needed because it's run as root in the container (which we can't avoid if we want to support Docker in docker)

On Windows, the same behavior is expected. The VM is using a network interface with a separate IP range. It would be great though if we could support this (e.g. by detecting the outer IP range) as this also breaks `balena scan` in WSL2.